### DISCUSSION TITLE
How to ensure scalability of these service? As workflow's num may grow, and the qps of one worflow may turn up

### DISCUSSION THREAD

**Discussion by yanghaogn** [Netflix/conductor#2805] [Q&A]
1. Where do workflow/task service、decider/queue service run ?
2. How to ensure scalability of these service? As workflow's num may grow, and the qps of one worflow may turn up
3. Does the output of each task store in Dynomite? And we can query the result of a workflowId's each task every time?

![image](https://user-images.githubusercontent.com/5836768/156318575-370a7024-b39e-48d2-a50e-0b131be0f7ea.png)

**Comment by v1r3n** [comment]
Conductor achieves scalability by scaling 4 core components of the system:

1. API Layer that is used by workers and users to start workflow executions
2. Queues that are used to schedule and maintain the tasks which are in progress
3. System task workers like HTTP, EVENT tasks that runs on Conductor server cluster
4. Persistence store that is used to store and index executions and workflow + task metadata

This is the general guidance on scaling each of these:

1. API layer is completely stateless and can be scaled horizontally - depending upon the hardware/VMs they are deployed, you can add more nodes to support the volume
2. Queues are backed by Dynomite/Redis and eventually are limited by the memory limit of Redis/Dynomite cluster.  You can vertically scale these services or shard them (if using Redis HA cluster or Dynomite)
3. System task workers are also stateless and can be scaled horizontally based on the load (we recommend running separate conductor cluster where `conductor.system-task-workers.enabled` property is set to `true` and keep this property set to `false` for the servers serving API requests in # 1 above when deployed in a very high throughput environment
4. Persistence store can be Dynomite - but it limits the amount of data you can store since its all in memory, or it can be something like Postgres, Mysql  or Cassandra.  Alternatively, you can remove and archive completed workflow to Elasticsearch and keep the Dynomite/Redis memory under the limits.

Happy to discuss further if you need more details on deployment strategies for highly scalable environment.

**Reply by yanghaogn** [reply]
Thanks for your reply. Another question about this
1. is the execution flow image from [Netflix conductor](https://www.slideshare.net/virenx/netflix-conductor) above old, or can you help supply a newest one?
2. If there are 100 millions workflows are executing at the same time, how to scale the  the decider service ?

**Reply by arketec** [reply]
I also have a question (or clarification). We noticed recently the time between tasks executing has slowed down. What used to be under a second between tasks (System tasks and/or SIMPLE tasks), now takes 10 secs or even 30 seconds in between tasks, while the actual tasks themselves run in a second or two.

Is this an issue with our dynomite setup? Right now, we run a 3 node cluster on a single VM, and we have left most of the configuration at the default levels.

we use redis as the backend and are running it in HA mode without a persistence layer, as we don’t need to persist the data.

edit:
here is a snip of what I mean
![conductor](https://user-images.githubusercontent.com/1336974/160024474-15e26be2-718f-4d90-97da-1ccf9bd4ac64.PNG)

**Reply by arketec** [reply]
Ok, I discovered why we were seeing these slowdowns. There was a bug in our conductor workers that was causing them to poll non-stop, and this is why we were seeing such huge slow downs because the server was getting overloaded by polling requests

**Reply by spasaria** [reply]
Hi @v1r3n I reread your answer several times trying to understand every single sentence.
Is it possible to discuss this topic further and in more detail? Maybe through a direct channel to surpass the github issue reply limits?
I sincerely hope you comply with my request for help 🙏

Thank you,
Andrea.

**Reply by aravindanr** [reply]
@spasaria you can use the Orkes slack channel, https://join.slack.com/t/orkes-conductor/shared_invite/zt-xyxqyseb-YZ3hwwAgHJH97bsrYRnSZg
